### DISCUSSION TITLE
Quick terminal tabs

### DISCUSSION THREAD

**Discussion by delicb** [ghostty-org/ghostty#2888] [Feature Requests, Ideas] [answered]
I know quick terminal does not support tabs, but I was wondering if it is a fundamental problem with having tabs in quick terminal or design decision or just "there are more important things to do now". All answers are ok, I am just looking for some insight. 

Context: All I really need is single Ghostty window, but need it on different Desktops (MacOS) because I am using terminal in couple of different contexts. Quick terminal is perfect fit for that. However, having only one tab is a bit limiting. 

Workarounds:
1. mess with window managers to move regular window to current Desktop and position it properly and hide it (simulate toggle). This is proving to be a bit difficult, apparently apple has made private some undocumented functions that automation tools (like BTT) were using.  
2. just make new Ghostty window in each Desktop and create some hotkeys for toggling its visibility - not sure how to approach toggle visibility thing now (but I am sure tools like BTT have some solution). Also - ideally I would love to have all commands, outputs, running TUIs "close", instead of having them scattered across different windows on different desktops. 
3. use splits instead of tabs - I am doing this now, but it is not ideal and serves a different purpose
4. use multiplexer tabs (e.g. tmux, zellij) inside quick terminal - this is viable option if there are no other options (at least in my case, I am not using multiplexer now)
5. hack around with Hammerspoon or similar tools to simulate quick terminal with regular Ghostty window

There are probably other workarounds (and I would love to hear them if anyone has suggestions). I would not want to sped a lot of time setting some of these workarounds and changing the muscle memory I have now if tabs in quick terminal are possible and will be coming. If not - that is fine as well, I will just adapt my approach (and this is where any input from people using similar flow is welcome).

**Comment by mitchellh** [comment, marked as answer]
It’s a somewhat fundamental problem: macOS doesn’t allow native tabs without a titlebar. So to support tabs we’d either have to add a titlebar to the window which looks kind of weird, or create a non native tabbing solution.

We’re already considering doing the latter (non native tabbing) to solve some other issues so my thinking was to just do that. But there’s no timeline on it.

**Reply by delicb** [reply]
Thanks for the response. I was half-expecting some fundamental problem behind it. 

I have my answer now - look for satisfiable alternative for now and keep an eye out in the future for potential in-app solutions.

**Reply by sassman** [reply]
For bottom anchored tabs a non native solution must be implemented anyways, mustn't it?

**Reply by hettiger** [reply]
> It’s a somewhat fundamental problem: macOS doesn’t allow native tabs without a titlebar. So to support tabs we’d either have to add a titlebar to the window which looks kind of weird, or create a non native tabbing solution.

Interesting, I'd actually prefer having a title bar. Was already wondering whether that's a possibility. Then, having native tabs over some non-native tabbing solution would also be nice. (Considering that you already use the native tabbing on the regular windows it makes even more sense to me.)

**Reply by merlinaudio** [reply]
+1 to just adding a title bar to the window. i'd be perfectly content with that - the only thing that would make it look weird is corner rounding, but can that not be disabled?

**Comment by pythoninthegrass** [comment]
>     4. use multiplexer tabs (e.g. tmux, zellij) inside quick terminal - this is viable option if there are no other options (at least in my case, I am not using multiplexer now)

This is what I was planning to do, but the [workaround](https://github.com/ghostty-org/ghostty/discussions/3483) for the default zellij keybinding for tabs isn't cooperating on my end. Quick terminal tabs and [broadcasting](https://github.com/ghostty-org/ghostty/discussions/3227) are the only blockers from making the switch to ghostty 🤞

**Comment by pluiedev** [comment]
See #5370 for a possible implementation

**Comment by abjugard** [comment]
My workflow regarding the terminal is basically that I exclusively use iTerm2:s hotkey window currently.

The main reason for this is that I use full screened apps quite a lot, so I'll have a few IDEs open in their spaces, a few browsers, some communications apps, and just switch between them with gestures or ctrl-arrows.

But I obviously still need access to the terminal though, hence quick terminals that pop down from the top of my monitor.

I generally have a bunch of tabs there, but rarely ever open a non-hotkey terminal window (maybe if I'm actively monitoring some logs I guess).

So as much as I would love to switch to Ghostty, it doesn't really work for me without tabs in quick terminal unfortunately.

Hope this feature eventually gets prioritised! 🤞🤞

**Reply by 00-kat** [reply]
I hope you read Mitchell's answer and #5370.

**Reply by abjugard** [reply]
I have, but it's been over a month with no progress there so I figured I'd just make a post here ¯\\\_(ツ)_/¯

**Comment by kirilldobr** [comment]
Really sad there is no tab option! Was looking for an alternative for Warp that has this feature, liked the UX implementation more than iTerm's (which doesn't focus on the input after you call for it, crazy), but this is a major limitation. Having tabs visually is perfectly fine UI-wise.

**Comment by uloco** [comment]
Adding an option to allow a titlebar in quick terminal would make me very happy and I don't think it should be big of an issue, no?

**Comment by icedice** [comment]
Tabs in quick terminal is pretty much whats keeping me from switching from Iterm2....

**Comment by 00-kat** [comment]
To everyone commenting here:
1. Mitchell has answered this discussion already; **please read it!** → https://github.com/ghostty-org/ghostty/discussions/2888#discussioncomment-11467352
2. This discussion has been closed. Our discussion rules ask that you don't revive old, closed, or outdated discussions → #6937.
3. This discussion is actually *very* outdated. Since its advent, *many* people have worked on getting tabs to work in the quick terminal: #5370, #8469, #9857. The last time a maintainer could spare the time to review the PRs, it was found to be very broken (see the second PR I linked). The last PR claims to have fixed all of them, but no maintainer with expertise in the area (there's surprisingly few) have had the time and willingness to review it yet.

**If you want to express support, or just really wish it was merged already**, I ask that, instead of commenting here, you instead:
1. Go to #9857, then **use reactions** to show your support:
   <img width="343" alt="Screenshot of hovering over the heart reaction." src="https://github.com/user-attachments/assets/85c0c09f-14c4-4abb-accb-63c7160cdd5e" />
   If you can't find it, these buttons are right at the bottom of the first comment. This is much more helpful than adding a comment—GitHub lets you sort by upvotes.
2. **Subscribe** to that issue, so that you can keep up with progress on this:
   <img width="340" alt="Screenshot of the subscribe button." src="https://github.com/user-attachments/assets/d08746ae-6ea6-48d2-9418-70c43ccc1956" />
   This one's at the bottom of the sidebar. If this is too noisy for you, or you don't care about progress, GitHub lets you select what you want to subscribe to: press the small “Customize” text that's above the subscribe button.

If you could not care less, and **want tabs in your quick terminal *right now***: you can build and **run #9857 yourself**! Many others are already doing this (https://github.com/ghostty-org/ghostty/pull/9857#issuecomment-4379360106), and haven't found any issues. Do keep in mind that it's effectively beta quality—but you could help squash those sneaky bugs left, like this kind person: https://github.com/ghostty-org/ghostty/pull/9857#issuecomment-4408050629. Instructions to build Ghostty are at https://ghostty.org/docs/install/build, and if you're stuck, those instructions don't work, or anything about it is confusing, please [open a new discussion] about it.

[open a new discussion]: https://github.com/ghostty-org/ghostty/discussions/new?category=q-a

If you're wondering **when maintainers are going to review those PRs**: please read https://github.com/ghostty-org/ghostty/pull/8469#issuecomment-3313176834 and (especially) https://github.com/ghostty-org/ghostty/pull/9857#issuecomment-4412039251.

But I ask that, whatever you do, you **do <ins>NOT</ins> add a “when will this be merged?”-style comment to that pull request**. If you want to review the code, feel free! And if you find a bug, go ahead and add a comment, and thank you. But adding more “this pull request is very useful for macOS users, because [long and well written argument], so mitchellh can you please review it?” or “I think it's best to merge this as-is, and fix bugs later” or “I can't wait for this PR to land! It's the only thing keeping me on iTerm2[^iterm2]” *isn't helpful at all*. All it does is send an email to *literal scores* of people subscribed to that PR—which might even include you.

And no: trying to get the maintainers' attention, or ask them to review it, isn't helpful. It happens literally every few weeks; the maintainers know of the pull request and will get to it when they're able to.

To prevent more necrobumps here, I'm going to lock this discussion[^upvotes].

[^iterm2]: FYI: I'm not picking on any specific comment when I say iTerm2; it's genuinely the one almost everyone has said they're stuck on!
[^upvotes]: I'll keep upvotes and reactions enabled so that you can express support here if you really want to.
